### PR TITLE
fix issue with invalid trades fees 

### LINKED ISSUE #4172: Disallow addition and editing of external trade that could result in invalid trade fees
## Problem Definition

At the moment it's possible to add an external trade through the API (or edit it) in a way that could result in invalid fees.

Specifically it seems that it's possible to have a fee but no fee currency which should not be possible.

We need to disallow this in the API and add a test for both add/edit that should fail.

## Task

The api code that adds a trade is here: https://github.com/rotki/rotki/blob/7f7478df40eb6b9397d3f0df945e23f0cc73a924/rotkehlchen/api/rest.py#L812

The api code that edits a trade is here:
https://github.com/rotki/rotki/blob/7f7478df40eb6b9397d3f0df945e23f0cc73a924/rotkehlchen/api/rest.py#L847

1. We need to make sure that if fee or fee_currency is given both of them are given and not only one. Also that fee is not ZERO.
2. Add tests for this.
3. Also add a changelog entry.
4. Add a DB upgrade to go through existing trades and if fee_currency is missing then also make fee as null.

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #4172

## Checklist
- [x] Added a changelog entry
- [x] Added a DB upgrade to go through existing trades and if fee_currency is missing then also make fee as null.
- [x] Updated schema  and typing to make `fee` and `fee_currency` required.
- [x] Also updated schema to check that `fee` is not zero.
- [x] Wrote tests

### CODE REVIEW

File: rotkehlchen/db/upgrades/v32_v33.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    """)
+
+
+def upgrade_v32_to_v33(db: 'DBHandler') -> None:
+    """Upgrades the DB from v32 to v33
+    - sets fee to null for existing trades if fee_currency is missing.
+    """
+    cursor = db.conn.cursor()
+    _update_fee_for_existing_trades(cursor)
+    db.conn.commit()
```
Comment (yabirgb) [inline rotkehlchen/db/upgrades/v32_v33.py]: Changes for the next version should go in `v31_v32.py`

**Review by yabirgb — COMMENTED** [review]
Hello @prettyirrelevant thanks! One first comment

File: docs/installation_guide.rst [hunk 2]
```diff
@@ -461,16 +461,19 @@ Before using `pip3`, let´s ensure we have the latest version::
...
-     ``$ LIBRARY_PATH=/usr/local/Cellar/sqlcipher/4.4.0/lib pip3 install pysqlcipher3``.
+    Make sure that pysqlcipher3 is properly installed. If ``$ pip3 freeze | grep pysqlcipher3`` returns nothing for you then it was not installed. Try to manually install only that dependency with the verbose option to see where it fails. ``$ pip3 install pysqlcipher3 -v``. If it fails at the stage of finding the library for ``-lsqlcipher`` or ``sqlcipher/sqlite3.h file not found`` then ``brew install sqlciper`` did not place the installed lib directory to the ``LIBRARY_PATH`` and you will have to do it manually.
```
Comment (LefterisJP) [inline docs/installation_guide.rst:472]: Why are all these changes appearing here again? You already made a PR for this. Is your PR based on latest develop?

File: rotkehlchen/api/rest.py [hunk 3]
```diff
@@ -854,8 +854,8 @@ def edit_trade(
             trade_type: TradeType,
             amount: AssetAmount,
             rate: Price,
-            fee: Optional[Fee],
-            fee_currency: Optional[Asset],
+            fee: Fee,
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:857]: all the typing changes in this file are not correct. `fee` and `fee_currency` are optional. What is not optional is having one but not the other.

File: rotkehlchen/api/v1/fields.py [hunk 4]
```diff
@@ -249,9 +249,8 @@ def _serialize(
             attr: str,  # pylint: disable=unused-argument
             obj: Any,  # pylint: disable=unused-argument
             **_kwargs: Any,
-    ) -> Optional[str]:
-        # Fee can be missing so we need to handle None when serializing from schema
-        return str(value) if value else None
+    ) -> str:
+        return str(value)
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/fields.py:253]: This is not correct. Fee can indeed be missing.

File: rotkehlchen/api/v1/fields.py [hunk 5]
```diff
@@ -374,9 +373,8 @@ def _serialize(
             attr: str,  # pylint: disable=unused-argument
             obj: Any,  # pylint: disable=unused-argument
             **_kwargs: Any,
-    ) -> Optional[str]:
-        # Asset can be missing so we need to handle None when serializing from schema
-        return str(value.identifier) if value else None
+    ) -> str:
+        return str(value.identifier)
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/fields.py:377]: What are you doing here? Why are you modifying the `AssetField` ?

File: rotkehlchen/api/v1/schemas.py [hunk 6]
```diff
@@ -513,11 +513,16 @@ class TradeSchema(Schema):
     trade_type = SerializableEnumField(enum_class=TradeType, required=True)
     amount = PositiveAmountField(required=True)
     rate = PriceField(required=True)
-    fee = FeeField(load_default=None)
-    fee_currency = AssetField(load_default=None)
+    fee = FeeField(required=True)
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:516]: no this is not what the issue asks:

> We need to make sure that if fee or fee_currency is given both of them are given and not only one. Also that fee is not ZERO.

File: rotkehlchen/db/upgrades/v32_v33.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+def _update_fee_for_existing_trades(cursor: 'Cursor') -> None:
+    cursor.execute("""
+    UPDATE trades
+    SET fee = NULL
+    WHERE fee_currency IS NULL
+    """)
+
+
+def upgrade_v32_to_v33(db: 'DBHandler') -> None:
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v32_v33.py:17]: The upgrade should be in the existing v31 to v32. No need to make a new one. A new DB upgrade is made only at release time.

**Review by LefterisJP — COMMENTED** [review]
Hmm @prettyirrelevant I think you misunderstood what the issue asks.

Please be in constant communication when working so you don't waste time in a wrong direction. Ask questions in discord. @yabirgb can help you if I am not there.

**Review by LefterisJP — COMMENTED** [review]
Looks much better. Still some small comments and one major for testing the upgrade.

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 8]
```diff
@@ -2485,6 +2486,23 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
             'SELECT COUNT(*) FROM sqlite_master WHERE type="table" AND name=?', (name,),
         ).fetchone()[0] == 0
 
+        db_v31.logout()
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2489]: What's this doing here? That db has already been logged out.

```suggestion
```

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 9]
```diff
@@ -2485,6 +2486,23 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
...
+
+    # Check that trades with fee missing sets fee_currency to NULL and vice versa
+    db = _init_db_with_target_version(
+        target_version=32,
+        user_data_dir=user_data_dir,
+        msg_aggregator=msg_aggregator,
+    )
+    cursor = db.conn.cursor()
+    trades = cursor.execute('SELECT * FROM trades WHERE id <> ? AND id <> ?', ("foo1", "foo2")).fetchall()  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2498]: please also in the beginning of the test assert that these trades are in the v31 DB with the mistaken fees.

Also don't use `<>` but `!=` as we do in the rest of the codebase.

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 10]
```diff
@@ -2485,6 +2486,23 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
...
+    assert trades == [
+        ("1111111", 1595640208, "external", "ETH", "USD", "buy", "1.5541", "22.1", "3.4", "USD", None, None),  # noqa: E501
+        ("1111112", 1595640208, "external", "ETH", "USD", "buy", "1.5541", "22.1", None, None, None, None),  # noqa: E501
+        ("1111113", 1595640208, "external", "ETH", "USD", "buy", "1.5541", "22.1", None, None, None, None),  # noqa: E501
+    ]
+    db_v31.logout()
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2504]: This is not doing anything.
```suggestion
```

**Review by LefterisJP — COMMENTED** [review]
More stuff.

File: rotkehlchen/tests/api/test_trades.py [hunk 11]
```diff
@@ -436,6 +436,57 @@ def test_add_trades(rotkehlchen_api_server):
...
+        'timestamp': 1595640208,
+        'location': 'external',
+        'base_asset': 'ETH',
+        'quote_asset': 'USD',
+        'trade_type': 'buy',
+        'amount': '1.5541',
+        'rate': '22.1',
+        'fee': '0.55',
+    }
+
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_trades.py:450]: nitpick: don't add unneeded lines
```suggestion
```

File: rotkehlchen/tests/api/test_trades.py [hunk 12]
```diff
@@ -436,6 +436,57 @@ def test_add_trades(rotkehlchen_api_server):
...
+    response = requests.put(
+        api_url_for(
+            rotkehlchen_api_server,
+            "tradesresource",
+        ), json=fee_and_no_fee_currency_trade,
+    )
+
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_trades.py:457]: nitpick: don't add unneeded lines. Part of the test that should be bundled together should have no newlines
```suggestion
```

File: rotkehlchen/tests/api/test_trades.py [hunk 13]
```diff
@@ -436,6 +436,57 @@ def test_add_trades(rotkehlchen_api_server):
...
+        'location': 'external',
+        'base_asset': 'ETH',
+        'quote_asset': 'USD',
+        'trade_type': 'buy',
+        'amount': '1.5541',
+        'rate': '22.1',
+        'fee': '0',
+        'fee_currency': 'USD',
+    }
+
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_trades.py:476]: nitpick: don't add unneeded lines. Part of the test that should be bundled together should have no newlines
```suggestion
```

File: rotkehlchen/tests/api/test_trades.py [hunk 14]
```diff
@@ -436,6 +436,57 @@ def test_add_trades(rotkehlchen_api_server):
...
+    response = requests.put(
+        api_url_for(
+            rotkehlchen_api_server,
+            "tradesresource",
+        ), json=fee_is_zero_trade,
+    )
+
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_trades.py:483]: nitpick: don't add unneeded lines. Part of the test that should be bundled together should have no newlines
```suggestion
```

File: rotkehlchen/api/v1/schemas.py [hunk 15]
```diff
@@ -517,6 +517,21 @@ class TradeSchema(Schema):
...
+    @validates_schema
+    def validate_fee_and_fee_currency(  # pylint: disable=no-self-use
+        self,
+        data: Dict[str, Any],
+        **_kwargs: Any,
+    ) -> None:
+        fee = data.get('fee')
+        fee_currency = data.get('fee_currency')
+
+        if not ((fee and fee_currency) or not (fee or fee_currency)):
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:529]: This condition is really complicated and hard to read. I had to write a test to make sure it's correct :sweat: 

Can you please simplify it?

```suggestion
        if not fee or not fee_currency and (fee or fee_currency):
```

Also perhaps explain in the docstring of the `validate_fee_and_fee_currency` function what it checks for.

**Review by LefterisJP — COMMENTED** [review]
Some more style and nitpick comments.

Also you will need to adjust the frontend integration tests as they now fail due to your change. Your change is correct though and the tests are wrong :smile: 

They were using `0` fee as input. The fix should be as simple as changing the fee from `0` to `1` for both trades in this json file: https://github.com/rotki/rotki/blob/develop/frontend/app/tests/e2e/fixtures/history/trades.json

File: rotkehlchen/api/v1/schemas.py (hunk 15 above)
Comment (prettyirrelevant) [reply rotkehlchen/api/v1/schemas.py]: So I replaced the logic with something more readable
```py
if not all([fee, fee_currency]) and any([fee, fee_currency]):
    ...
```

Although the logic operator was faster, this is supposedly more readable and easier to understand.
- Using logic operators only  --->`0.035516125`
- Using `any` and `all` ---> `0.201628292`

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 16]
```diff
@@ -2461,7 +2462,14 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
         'manually_tracked_balances;',
     ).fetchall()
 
-    db_v31.logout()
+    # check that the trades with invalid fee/fee_currency are present at this point
+    trades_before = cursor.execute('SELECT * FROM trades WHERE id != ? AND id != ?', ("foo1", "foo2")).fetchall()  # noqa: E501
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/tests/db/test_db_upgrades.py:2466]: You actually still use double quotes here :smiling_face_with_tear:

**Review by LefterisJP — APPROVED** [review]
Thanks @prettyirrelevant. Looks better now. I will merge. Backend tests passed. The frontend e2e tests still fail. But to be honest I don't know why. Making the change you did they pass locally for me. I will ask @lukicenturi or @kelsos to look at this test in another PR. This way only frontend tests and code will be affected in the other PR.

File: rotkehlchen/api/v1/schemas.py (hunk 15 above)
Comment (LefterisJP) [reply rotkehlchen/api/v1/schemas.py]: Normally I would go for speed especially with such a big difference of almost 6x. But we can adjust if this becomes a bottleneck.

Adding the comment already helps a lot with readability.

**Pull request merged**